Add some documentation about the official nix formatter (#1129)
NixOS/nix.dev: source/guides/recipes/autoformatting.md

(autoformatting)=
# Autoformatting

[`nixfmt`](https://github.com/NixOS/nixfmt) is the official Nix autoformatter.

It is not yet wired up to any official tooling. Subscribe to [PR
11252 to NixOS/nix](https://github.com/NixOS/nix/pull/11252) for updates on
that effort.

The easiest way to leverage `nixfmt` in your project is with the `nixfmt-tree`
wrapper. Here's how:

- For `nix fmt` to format all Nix files, add this to the `flake.nix` outputs:

  ```nix
  formatter.''${system} = nixpkgs.legacyPackages.''${system}.nixfmt-tree;
  ```

> - The same can be done more efficiently with the `treefmt` command,

Reviewer: Please show exactly one way of doing things, and since official documentation is committed to stable features (and you already point out that it's more efficient) let's use the one where you add the package to your shell environment.

Reviewer 2: > Please show exactly one way of doing things

I'm not sure I agree we should limit ourselves to only documenting a single approach to things.

`nix fmt` and `nix-shell --run treefmt` both have their own pros and cons, I'd expect users would want to be aware of each.

Reviewer: Realty is, we have to balance cognitive load against factual completeness. This place is about managing cognitive load, reference documentation is for completeness. I'd appreciate  discussion in reference documentation and issue threads on the costs and benefits of various approaches, but the motto of nix.dev is "getting things done" -- not "agonizing over the options". We're free to agonize in these behind the scenes exchanges, but we must not spill that on beginners.

Reviewer 3: Tbh I'm not sure if `nix fmt` has a benefit other than it being a shorter command, but we should recommend `direnv` anyways, in which case `treefmt` has the exact same length. I'd be fine with dropping `nix fmt`, though open about other benefits I'm missing!

  which you can get in `nix-shell`/`nix develop` by extending `mkShell` using

  ```nix
  mkShell {
    packages = [ pkgs.nixfmt-tree ];
  }
  ```

  You can then also use `treefmt` in a pre-commit/pre-push [Git hook](https://git-scm.com/docs/githooks)
  and `nixfmt` with your editors format-on-save feature.

- To check formatting in CI, run the following in a checkout of your Git repository:
  ```
  treefmt --ci
  ```

For more flexibility, you can customize this package using
```nix
nixfmt-tree.override {
  settings = { /* treefmt config */ };
  runtimePackages = [ /* List any formatters here */ ];
}
```

> Alternatively you can switch to the more fully-featured [treefmt-nix](https://github.com/numtide/treefmt-nix).

Reviewer: Rather than propose ever more alternative ways of doing things, I'd appreciate (and review) a guide to set up a pre-commit hook with treefmt-nix.

Reviewer 2: While they both relate to formatting, and both produce a configured instance of treefmt, `pkgs.nixfmt-tree`, `pkgs.treefmt.withConfig`, and `treefmt-nix` all exist for different reasons and suit different use cases.

Users who _just_ want to configure formatting nix files in their project shouldn't need to download or configure treefmt-nix; there's a solution built into nixpkgs.

Users who want a more complex formatting setup, where they end up configuring several treefmt options, may benefit from using either treefmt-nix or `pkgs.treefmt.withConfig`.

Reviewer: Right, and then we'd need to ask what's the purpose of any given article, and whether this information may be better placed elsewhere, such as is Nixpkgs contributor docs.

This sort of problem appears all of the time here, and sometimes one first needs to write an entirely guide to to write the guide one originally wanted to write. Otherwise it becomes just as unfocused and confusing as we still have an unfortunate reputation for.

(autoformatting)=
# Autoformatting

[`nixfmt`](https://github.com/NixOS/nixfmt) is the official Nix autoformatter.

> It is not yet wired up to any official tooling. Subscribe to

Reviewer 2: Very minor, but the wording feels backwards to me... The onus should be on the tooling to wire up; it is out of nixfmt's hands what tooling decides to use it.

I'm also struggling to phrase it, but maybe:

```suggestion
Official tooling has yet to be integrated with nixfmt. Subscribe to
```

or:

```suggestion
Official tooling currently doesn't use nixfmt out of the box. Subscribe to
```

[NixOS/nix PR #11252](https://github.com/NixOS/nix/pull/11252) for updates on that effort.

Because `nixfmt` doesn't support formatting whole directory trees, you need
additional tooling such as `treefmt`. The `nixfmt-tree` package provides a
`treefmt` pre-configured to run `nixfmt` on all nix files in your project. Just
add it to your shell:

```nix
mkShell {
  packages = [ pkgs.nixfmt-tree ];
}
```

Note: this assumes you're working in a git repository with a [`.git/index` file
at the

> root](https://github.com/NixOS/nixpkgs/blob/bbcdf758d61b6095693d00dea8a75061bf89b264/pkgs/by-name/ni/nixfmt-tree/package.nix#L33-L34).

Reviewer 2: :+1: for mentioning, however while it's currently correct, it will soon be outdated.

https://github.com/numtide/treefmt/pull/571 changed treefmt's default behaviour, so now it will dynamically ask `git` for the root of the workspace.

With that change, `tree-root-file` is no longer needed in most cases. Therefore, `nixfmt-tree` will remove its `tree-root-file = ".git/index"` definition once the next treefmt version is released.

It will _still_ be correct to say something like:

> this assumes you're project is in a git repository, and you wish to treat the entire repo as your project to be formatted

, however it will no longer be necessary to mention `.git/index` specifically or the implications that has for git worktrees.

> If you need to change the root detection, or configure treefmt to format other (non-nix)

Reviewer 2: Maybe better to be more general and less specific regarding treefmt options?

```suggestion
If you need to configure any [treefmt options], or enable formatting other (non-nix)
```

We can also use this as an opportunity to link to the treefmt config docs:

```md
[treefmt options]: https://treefmt.com/latest/getting-started/configure/#global-options
```

files, you can use `treefmt.withConfig`:

```nix
pkgs.treefmt.withConfig {
  runtimeInputs = [
    pkgs.nixfmt-rfc-style
    pkgs.ruff
  ];

  settings = {
    # Customize detection of the root of the project.
    tree-root-file = "flake.nix";

    # Configure nixfmt for .nix files.
    formatter.nixfmt = {
      command = "nixfmt";
      includes = [ "*.nix" ];
    };

    # And for .py file.
    formatter.ruff = {
      command = "ruff";
      options = [ "format" ];
      includes = [ "*.py" ];
    };
  };
}
```

This can get a little tedious.
[treefmt-nix](https://github.com/numtide/treefmt-nix) has a big library of

> preconfigured formatters, and provides a `check` derivation you can use in CI.

Reviewer 2: If we're mentioning the `check` derivation, should we include a (simple) example of using it?

If flakes are acceptable in nix.dev guides, maybe adding the drv to a flake's `checks` output would be a suitable example?

Such an example may be a good opportunity to "sneak in" defining `formatter` and `devshell` outputs, although not if that ends up making the example less simple.

Author: I agree it would be nice to include a simple example of using the `check` derivation, but that requires either using npins or flakes to pull in treefmt-nix. It gets messy kind of fast, and I'm lightly leaning towards "just send them to treefmt-nix to figure this out", as I've done here.

I'm very ambivalent on this, though. I'm going to leave this unresolved, perhaps someone with strong opinions will show up.
